src/frontend/flight/sync.rs

```diff
@@ -266,89 +271,113 @@ impl SeafowlDataSyncManager {
-            let orseq = entry
-                .syncs
-                .iter()
-                .map(|s| (s.origin, s.sequence_number))
-                .unique()
-                .collect::<Vec<_>>();
-            self.remove_sequence_locations(url.clone(), orseq);
+        let last_sequence_number = entry.syncs.last().unwrap().sequence_number;
+
+        let mut table = DeltaTable::new(log_store.clone(), Default::default());
+        table.load().await?;
+
+        if let Some(table_seq) = self.table_sequence(&table).await?
+            && table_seq > last_sequence_number
+        {
+            // TODO 1: partial skipping if only a subset of syncs older than the commited sequence
+            // TODO 2: persist the final flag to to enable >= comparison
+            info!(
+                "Location at {url} already durable up to {table_seq}, skipping {}",
+                last_sequence_number
+            );
             self.remove_sync(&url);
-            self.advance_durable();
+            return Ok(());
+        }
+
+        // Use the schema from the object store as a source of truth, since it's not guaranteed
+        // that any of the entries has the full column list.
+        let full_schema = TableProvider::schema(&table);
+
+        // Generate a qualifier expression for pruning partition files and filtering the base scan
+        let qualifier = self.construct_qualifier(full_schema.clone(), entry)?;
+
+        // Iterate through all syncs for this table and construct a full plan by applying each
+        // individual sync
+        let base_plan = LogicalPlanBuilder::scan_with_filters(
+            SYNC_REF,
+            provider_as_source(Arc::new(table.clone())),
+            None,
+            vec![qualifier.clone()],
+        )?
+        .build()?;
+        let mut sync_df = DataFrame::new(self.context.inner.state(), base_plan);
+
+        for sync in &entry.syncs {
+            let data = sync.batches.clone();
+            sync_df =
+                self.apply_sync(&entry.pk_columns, full_schema.clone(), sync_df, data)?;
         }
 
+        let input_plan = sync_df.create_physical_plan().await?;
+
+        // To exploit fast data upload to local FS, i.e. simply move the partition files
+        // once written to the disk, try to infer whether the location is a local dir
+        let local_data_dir = if url.starts_with("file://") {
+            Some(log_store.root_uri())
+        } else {
+            None
+        };
+
+        // Dump the batches to the object store
+        let adds = plan_to_object_store(
+            &self.context.inner.state(),
+            &input_plan,
+            log_store.object_store(),
+            local_data_dir,
+            self.context.config.misc.max_partition_size,
+        )
+        .await?;
+
+        let mut actions: Vec<Action> = adds.into_iter().map(Action::Add).collect();
+
+        // Prune away files that are refuted by the qualifier
```

Comment:
as in partitions that we overwrote, right?

Reply:
Yup, correct.